source/slang/slang-lower-to-ir.cpp

```diff
@@ -6813,7 +6813,22 @@ struct StmtLoweringVisitor : StmtVisitor<StmtLoweringVisitor>
             IRBuilder subBuilder = *getBuilder();
             subBuilder.setInsertInto(info->initialBlock);
             subContext.irBuilder = &subBuilder;
-            auto caseVal = getSimpleVal(context, lowerRValueExpr(&subContext, caseStmt->expr));
+
+            auto caseValInfo = lowerRValueExpr(&subContext, caseStmt->expr);
+            if (caseValInfo.flavor != LoweredValInfo::Flavor::Simple)
+            {
+                // This expression would emit instructions other than a direct
+                // value, which we don't support yet here. If there was a
+                // constant value found during checking, we can use that value
+                // instead.
+                auto constVal = as<ConstantIntVal>(caseStmt->exprVal);
```

Comment:
Should it always be the case that the const Val exists? If not this should have been a  semantic error, right?

Reply:
I thought there could be corner cases where a constant value may not be known at compile-time, like specialization constants in Vulkan. However, it appears that these don't work in switch cases yet anyway. So as it stands, I think `constVal` should always exist, with a diagnostic emitted from `slang-check-stmt.cpp` if not.

Reply:
Yeah, I think we should just emit using the Val instead.

Reply:
Just make sure when we check a switch case, we are folding the case value using CompileTime kind instead of LinkTime or SpecConst time.

Reply:
If we always just use `exprVal` instead of `expr`, wouldn't that mean that the names of the constant variables would no longer be preserved in the switch cases if emitting into a high-level language like GLSL? I'm personally completely fine with that though.

Reply:
It appears that the `CompileTime` kind is already being used (via `checkConstantIntVal`), so that's good.

Reply:
Yeah we never preserve the names, so this change won’t make things worse.